lib/Doctrine/DBAL/Connections/PrimaryReplicaConnection.php

```diff
@@ -0,0 +1,359 @@
+ * operations only.
+ *
+ * This connection is limited to replica operations using the
+ * Connection#executeQuery operation only, because it wouldn't be compatible
+ * with the ORM or SchemaManager code otherwise. Both use all the other
+ * operations in a context where writes could happen to a replica, which makes
+ * this restricted approach necessary.
+ *
+ * You can manually connect to the primary at any time by calling:
+ *
+ *      $conn->connect('primary');
+ *
+ * Instantiation through the DriverManager looks like:
+ *
+ * @example
+ *
+ * $conn = DriverManager::getConnection(array(
+ *    'wrapperClass' => 'Doctrine\DBAL\Connections\PrimaryReplicaConnection',
+ *    'driver' => 'pdo_mysql',
+ *    'primary' => array('user' => '', 'password' => '', 'host' => '', 'dbname' => ''),
+ *    'replica' => array(
+ *        array('user' => 'replica1', 'password', 'host' => '', 'dbname' => ''),
+ *        array('user' => 'replica2', 'password', 'host' => '', 'dbname' => ''),
+ *    )
+ * ));
+ *
+ * You can also pass 'driverOptions' and any other documented option to each of this drivers to pass additional information.
+ */
+final class PrimaryReplicaConnection extends Connection
+{
+    /**
+     * Primary and Replica connection (one of the randomly picked replicas).
+     *
+     * @var DriverConnection[]|null[]
+     */
+    private $connections = ['primary' => null, 'replica' => null];
+
+    /**
+     * You can keep the replica connection and then switch back to it
+     * during the request if you know what you are doing.
+     *
+     * @var bool
+     */
+    private $keepReplica = false;
+
+    /**
+     * Creates Primary Replica Connection.
+     *
+     * @param mixed[] $params
+     *
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        array $params,
+        Driver $driver,
+        ?Configuration $config = null,
+        ?EventManager $eventManager = null
+    ) {
+        if (! isset($params['replica'], $params['primary'])) {
+            throw new InvalidArgumentException('primary or replica configuration missing');
+        }
+
+        if (count($params['replica']) === 0) {
+            throw new InvalidArgumentException('You have to configure at least one replica.');
+        }
+
+        $params['primary']['driver'] = $params['driver'];
+        foreach ($params['replica'] as $replicaKey => $replica) {
+            $params['replica'][$replicaKey]['driver'] = $params['driver'];
+        }
+
+        $this->keepReplica = (bool) ($params['keepReplica'] ?? false);
+
+        parent::__construct($params, $driver, $config, $eventManager);
+    }
+
+    /**
+     * Checks if the connection is currently towards the primary or not.
+     */
+    public function isConnectedToPrimary(): bool
+    {
+        return $this->_conn !== null && $this->_conn === $this->connections['primary'];
+    }
+
+    public function connect(?string $connectionName = null): bool
```

Comment:
Please see https://github.com/doctrine/dbal/pull/4054#pullrequestreview-425775128. I am concerned about using a proper class name and leaving all downsides of the old API. That means that if we decide to fix this downsides in a BC-breaking way, we'll have to use another class name. It's better to try an fix both the class name and the API problem in one shot.

Reply:
> It changes the semantic of the connect() method by adding an optional argument.

https://github.com/doctrine/dbal/blob/0ad4c165318128c50d9a425c8c15520c094f465b/lib/Doctrine/DBAL/Connections/MasterSlaveConnection.php#L128

There's no new argument, the only existing argument was already optional, wasn't it?

Reply:
Oh right you were talking about the one in the interface, I get it now… how do we fix this?

Reply:
Something like that. I haven't use this component, so I don't know what the intended use is. E.g. why does a single connection need to be stateful and switch between different underlying connection? If a given component is about to make a query or execute a statement, how does it know if it needs to switch or the connection is already switched to the proper endpoint?

Reply:
@beberlei @lstrojny might be able to help with this?

The more I read about this, the less I see why you would need to be able to `$this->connect('replica')`, since `replica` is the default, and since `If Primary was picked once during the lifetime of the connection it will always get picked afterwards.`

> If a given component is about to make a query or execute a statement, how does it know if it needs to switch or the connection is already switched to the proper endpoint?

It could use `isConnectedToPrimary()` for that, but apparently it shouldn't even need to, since `Primary picked when 'exec', 'executeUpdate', 'insert', 'delete', 'update', 'createSavepoint', 'releaseSavepoint', 'beginTransaction', 'rollback', 'commit', 'query' or l'prepare' is called.`

So the only use case I see for `$this->connect('primary')` is when you know you are going to perform a call to `executeQuery` and then calls that require using the primary connection, and you want to make sure you only ever connect to one things for performance reasons.

Reply:
I see two ways of implementing the primary/replica connection consumer in an application:

1. The actual endpoint being used for a specific statement is abstracted out and is chosen based on whether the statement is read or write. In this case, the application should depend on a `Connection` interface (no additional methods exposed).
2. The application wants to control which endpoint to use for a specific statement. This way, it's given two or more independent connections that don't expose any additional methods either.

All this `$this->connect('primary' /* ATTENTION, MAGIK STRING LITERAL! */)` and `$this->forceUsingPrimary(/* ATTENTION, MOAR MAGIK! */)` looks like devilry to me.

Reply:
With 1. we would drop the argument and `isConnectedToPrimary()`, so I could restore `final`, right? I love that keyword, it usually means fewer headaches with BC-breaks, that why I tend to use it (maybe a bit too much, as you pointed out).
2. This could be fixed by moving this grouping of connections as a concern for a DIC. The doctrine bundle would name connections according to a pattern, like `connection_foo_primary`, `connection_foo_replica0`, `connection_foo_replica1`… If that's not convenient enough, a service locator could be created to have `$container->get('primary')`, `$connection->get('replica0')`, etc. (where `$container` is a service locator that contains only these 3 connections). 

I'm going to push another commit showing how this could look like. If we do that, I think we should also drop `$keepReplica`: 


```
    /**
     * You can keep the replica connection and then switch back to it
     * during the request if you know what you are doing.
     *
     * @var bool
     */
    private $keepReplica = false;
```

Reply:
There is going to be an issue with the tests though: they heavily rely on `isConnectedToPrimary`

Reply:
It could be tested properly if this was a wrapper over multiple other connections that could be mocked. The "is connected to" is an implementation detail, not the behavior that needs to be tested.

Reply:
> Or, looking from another angle, where does the requirement of being connected only to one endpoint at a time comes from?

I'd say the belief that performance will be bad or results will be inaccurate if you perform a write on the primary db and immediately read from the replica. I have no idea if this holds true or not.

Reply:
Disclaimer: it’s more than 10 years that I did work on the original implementation.

The original design premise was "read from a replication, write to primary". This is fine until the first write, then we need to assume the replication to be out of sync, since we cannot control the specific replication algorithm from a database abstraction perspective.

I am nowadays of the opinion of a) this code should have never made it into Doctrine and b) what a seperate project (which may or may not live inside of Doctrine) could provide is building blocks for good multi-connection support. Typically the topologies of replicated or distributed setups are quite application specific, as are the specific configurations and therefore guarantees so providing flexible building blocks over a single implementation, albeit "conservative" in it’s choices, would be a better way.

Reply:
Here you go, @beberlei. Thank you for the confirmation, @lstrojny.